packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,238 @@
+    let resolvedConditionType
+
+    return await this.atomicPhase_(
+      async (manager: EntityManager) => {
+        resolvedConditionType =
+          DiscountConditionService.resolveConditionType_(data)
+
+        if (!resolvedConditionType) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Missing one of products, collections, tags, types or customer groups in data`
+          )
+        }
+
+        const discountRepo: DiscountRepository = manager.getCustomRepository(
+          this.discountRepository_
+        )
+
+        let discount
+        // in case id is passed, we retrieve the discount
+        if (typeof discountIdOrDiscount === `string`) {
+          discount = await discountRepo.findOne({
+            where: { id: discountIdOrDiscount },
+          })
+
+          if (!discount) {
+            throw new MedusaError(
+              MedusaError.Types.NOT_FOUND,
+              `Discount with ${discountIdOrDiscount} was not found`
+            )
+          }
+        } else {
+          discount = discountIdOrDiscount
+
+          // ensure discount has rule_id, will also ensure that object is in fact discount
+          if (!discount?.rule_id) {
+            throw new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              `Discount is missing rule_id`
+            )
+          }
+        }
+
+        const discountConditionRepo: DiscountConditionRepository =
+          manager.getCustomRepository(this.discountConditionRepository_)
+
+        if (data.id) {
+          // throws if condition or discount does not exist
+          const condition = await this.retrieve(data.id, discount.id)
+
+          return await discountConditionRepo.addConditionResources(
+            condition.id,
+            resolvedConditionType.resource_ids,
+            resolvedConditionType.type,
+            true
+          )
+        }
+
+        const created = discountConditionRepo.create({
+          discount_rule_id: discount.rule_id,
+          operator: data.operator,
+          type: resolvedConditionType.type,
+        })
+
+        const discountCondition = await discountConditionRepo.save(created)
+
+        return await discountConditionRepo.addConditionResources(
+          discountCondition.id,
+          resolvedConditionType.resource_ids,
+          resolvedConditionType.type
+        )
+      },
+      async (err: { code: string }) => {
+        if (err.code === PostgresError.DUPLICATE_ERROR) {
+          // A unique key constraint failed meaning the combination of
+          // discount rule id, type, and operator already exists in the db.
+          throw new MedusaError(
+            MedusaError.Types.DUPLICATE_ERROR,
+            `Discount Condition with operator '${data.operator}' and type '${resolvedConditionType?.type}' already exist on a Discount Rule`
+          )
+        }
+      }
+    )
+  }
+
+  async remove(discountConditionId: string): Promise<void> {
```

Comment:
the return type of the remove method from a repository is the entity, either we do not `return await remove` at the end or we change the return type.